clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -446,56 +485,47 @@ public int compare(ProducerBatch o1, ProducerBatch o2) {
      *         RecordBatch.NO_SEQUENCE.
      */
     synchronized int firstInFlightSequence(TopicPartition topicPartition) {
-        PriorityQueue<ProducerBatch> inFlightBatches = inflightBatchesBySequence.get(topicPartition);
-        if (inFlightBatches == null)
+        if (!hasInflightBatches(topicPartition))
             return RecordBatch.NO_SEQUENCE;
 
-        ProducerBatch firstInFlightBatch = inFlightBatches.peek();
-        if (firstInFlightBatch == null)
+        ProducerBatch first = topicPartitionBookkeeper.get(topicPartition).inflightBatchesBySequenceNumber.peek();
+        if (first == null)
             return RecordBatch.NO_SEQUENCE;
 
-        return firstInFlightBatch.baseSequence();
+        return first.baseSequence();
     }
 
     synchronized ProducerBatch nextBatchBySequence(TopicPartition topicPartition) {
-        PriorityQueue<ProducerBatch> queue = inflightBatchesBySequence.get(topicPartition);
-        if (queue == null)
-            return null;
+        PriorityQueue<ProducerBatch> queue = topicPartitionBookkeeper.get(topicPartition).inflightBatchesBySequenceNumber;
         return queue.peek();
     }
 
     synchronized void removeInFlightBatch(ProducerBatch batch) {
-        PriorityQueue<ProducerBatch> queue = inflightBatchesBySequence.get(batch.topicPartition);
-        if (queue == null)
-            return;
-        queue.remove(batch);
+        if (hasInflightBatches(batch.topicPartition)) {
+            PriorityQueue<ProducerBatch> queue = topicPartitionBookkeeper.get(batch.topicPartition).inflightBatchesBySequenceNumber;
+            queue.remove(batch);
+        }
     }
 
     synchronized void maybeUpdateLastAckedSequence(TopicPartition topicPartition, int sequence) {
         if (sequence > lastAckedSequence(topicPartition))
-            lastAckedSequence.put(topicPartition, sequence);
+            topicPartitionBookkeeper.get(topicPartition).lastAckedSequenceNumber = sequence;
     }
 
     synchronized int lastAckedSequence(TopicPartition topicPartition) {
```

Comment:
By the way, do you think it would be safer to return `Optional<Integer>` and keep the use of the sentinel confined to this class. Same thing for `lastAckedOffset`.

Reply:
Yea, I think it'd be good to do this. I also had this in mind but since the Java code in Kafka doesn't really use these constructs, I left them. Anyway I'll try it out.

Reply:
Also, by "this class" did you mean TransactionManager or TopicPartitionBookkeeper? I assumed the former.